pkg/scheduler/util/scheduler_helper.go

```diff
@@ -55,53 +60,113 @@ func PredicateNodes(task *api.TaskInfo, nodes []*api.NodeInfo, fn api.PredicateF
+				if err != nil {
+					appendError(err)
+				}
+			}(i, priorityConfig)
+		} else {
+			results[i] = make(schedulerapi.HostPriorityList, len(nodes))
+		}
+	}
+	processNode := func(index int) {
+		nodeInfo := nodeNameToInfo[nodes[index].Name]
+		var err error
+		for i := range priorityConfigs {
+			if priorityConfigs[i].Function != nil {
+				continue
+			}
+			results[i][index], err = priorityConfigs[i].Map(task.Pod, nil, nodeInfo)
+			if err != nil {
+				appendError(err)
+				return
+			}
 		}
+	}
+	workqueue.Parallelize(16, len(nodes), processNode)
+	for i, priorityConfig := range priorityConfigs {
+		if priorityConfig.Reduce == nil {
+			continue
+		}
+		wg.Add(1)
+		go func(index int, config algorithm.PriorityConfig) {
+			defer wg.Done()
+			if err := config.Reduce(task.Pod, nil, nodeNameToInfo, results[index]); err != nil {
+				appendError(err)
+			}
+			if glog.V(10) {
+				for _, hostPriority := range results[index] {
+					glog.Infof("%v -> %v: %v, Score: (%d)", task.Pod.Name, hostPriority.Host, config.Name, hostPriority.Score)
+				}
+			}
+		}(i, priorityConfig)
+	}
+	// Wait for all computations to be finished.
+	wg.Wait()
+	if len(errs) != 0 {
+		return schedulerapi.HostPriorityList{}, errors.NewAggregate(errs)
+	}
 
-		workerLock.Lock()
-		nodeScores[score] = append(nodeScores[score], node)
-		workerLock.Unlock()
+	// Summarize all scores.
+	result := make(schedulerapi.HostPriorityList, 0, len(nodes))
+	for i := range nodes {
+		result = append(result, schedulerapi.HostPriority{Host: nodes[i].Name, Score: 0})
+		for j := range priorityConfigs {
+			result[i].Score += results[j][i].Score * priorityConfigs[j].Weight
+		}
 	}
-	workqueue.ParallelizeUntil(context.TODO(), 16, len(nodes), scoreNode)
-	return nodeScores
+
+	return result, nil
 }
 
 // SortNodes returns nodes by order of score
-func SortNodes(nodeScores map[float64][]*api.NodeInfo) []*api.NodeInfo {
+func SortNodes(priorityList schedulerapi.HostPriorityList, nodesInfo map[string]*api.NodeInfo) []*api.NodeInfo {
 	var nodesInorder []*api.NodeInfo
-	var keys []float64
-	for key := range nodeScores {
-		keys = append(keys, key)
-	}
-	sort.Sort(sort.Reverse(sort.Float64Slice(keys)))
-	for _, key := range keys {
-		nodes := nodeScores[key]
-		nodesInorder = append(nodesInorder, nodes...)
+
+	sort.Sort(sort.Reverse(priorityList))
+
+	for _, hostPriority := range priorityList {
+		node := nodesInfo[hostPriority.Host]
+		nodesInorder = append(nodesInorder, node)
 	}
+
 	return nodesInorder
 }
 
 // SelectBestNode returns best node whose score is highest, pick one randomly if there are many nodes with same score.
-func SelectBestNode(nodeScores map[float64][]*api.NodeInfo) *api.NodeInfo {
-	var bestNodes []*api.NodeInfo
-	maxScore := -1.0
-	for score, nodes := range nodeScores {
-		if score > maxScore {
-			maxScore = score
-			bestNodes = nodes
-		}
-	}
+func SelectBestNode(priorityList schedulerapi.HostPriorityList) string {
+	sort.Sort(sort.Reverse(priorityList))
```

Comment:
No need to sort them, its time complexity is `O(NlogN)`, previous is `O(N)`.

Reply:
@hex108 We need sort the HostPriorityList and select the max score node at random

Reply:
We just need find MaxScores nodes, and pick one of them randomly.

Reply:
Done